### ISSUE TITLE
Subscribe.attributeChanged() calls _updateLinks even in idle

### ISSUE THREAD

**Issue by cxbrooks** [cxbrooks/test#54]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#57
From: @cxbrooks
Reported version: 6.0.1
CC:   pt-dev@chess.eecs.berkeley.edu

**Comment by cxbrooks** [comment]
From the Rome 1/17 telcon:

Subscriber: attributeChanged() calls _updateLinks() even if
the model is not running.  _updateLinks() calls _findPublisher().
We should only do this if the model is running by checking the Manager.

This causes problems with large models.

**Comment by cxbrooks** [comment]
I updated Subscriber.attributeChanged() so that we only call
_updateLinks() if the Manager says we are idle.  Otherwise,
we call _updateLinks() in preinitialize().

**Comment by cxbrooks** [comment]
Edward's 2007/03/27 email:
--start--
I've checked in a couple more performance improvements with
shared parameters.

Fundamentally, whenever one of these is created,
it must traverse the model to find at least one other shared parameter
from which to infer its value.  Since it only needs to find one, this
operation is now constant time (it does not grow with model size).
Previously, the code would find _all_ shared parameters, which is an
order N operation, where N is the model size. Thus, if there were M
shared parameters, we were getting N*M operations during model
construction.

When the value of a shared parameter is set (which also occurs during
model construction if you are not using default values), then we need
to again traverse the model to find all other shared parameters and
set their values.  This is an order N operation, so again we have
N*M complexity.  I have changed things so that this traversal is not
performed if the value is the same as what has already been inferred
from the context. In this case, we can assume that all other instances
of the shared parameter have the same value.

These two changes reduce the complexity during model construction
to order N, which should significantly improve both opening of models
and interactivity when editing models.

I'm going to now take a look at the preinitialize performance problems.
--end--

Edward's change to Subscriber.attributeChanged():

revision 1.18
date: 2007/03/27 17:20:52;  author: eal;  state: Exp;  lines: +14 -42
Update links in attributeChanged() and not in setContainer().
                       
 // NOTE: There used to be some logic here to call
 // _updateLinks() only if the manager indicates we
 // are running, and otherwise to set _updatedLinks
 // to false. This is no longer necessary as
 // the attributeChanged() method is called at
 // appropriate times, and tests show it is called
 // exactly once per publisher. Moreover, it really
 // isn't correct to defer this, as _updateLinks()
 // handles connectivity information. If we want to,
 // for example, highlight dependents before a model
 // has been run, this has to be done here.
